[clang] Define ptrauth_sign_constant builtin.

ahmedbougacha opened the pull request:
This is a constant-expression equivalent to
ptrauth_sign_unauthenticated.  Its constant nature lets us guarantee
a non-attackable sequence is generated, unlike
ptrauth_sign_unauthenticated which we generally discourage using.*

It being a constant also allows its usage in global initializers, though
requiring constant pointers and discriminators.

ahmedbougacha committed:
[clang] Define ptrauth_sign_constant builtin.

This is constant-expression equivalent to __builtin_ptrauth_sign,
allowing its usage in global initializers, but requiring constant
pointers and discriminators.

Co-Authored-By: John McCall <rjmccall@apple.com>

efriedma-quic commented:
Why do we want a separate builtin, as opposed to just constant-folding calls to __builtin_ptrauth_sign?

ahmedbougacha commented:
> Why do we want a separate builtin, as opposed to just constant-folding calls to __builtin_ptrauth_sign?

That's a good question.  Mechanically, I assumed constant-evaluating `__builtin_ptrauth_sign_unauthenticated` would be a challenge, but looking around it doesn't seem particularly difficult.

Conceptually though, the distinction does seem useful, because we treat `__builtin_ptrauth_sign_unauthenticated` as dangerous (because in general we can't guarantee that it won't get lowered to a signing oracle), and only to be used carefully, when absolutely necessary (e.g., in a dynamic loader.)
`__builtin_ptrauth_sign_constant` is safe, because we do always have safe ways to materialize signed constants (with the constants in IR, and either relocations or safe pseudos in the backend.)   (but cc @rjmccall @ahatanak for the real clang perspective ;)

ahmedbougacha committed:
Document ptrauth_sign_constant in clang/docs/PointerAuthentication.

kovdan01 commented on clang/include/clang/CodeGen/CodeGenABITypes.h:
```diff
@@ -104,6 +104,12 @@ llvm::Type *convertTypeForMemory(CodeGenModule &CGM, QualType T);
 
+/// Return a signed constant pointer.
+llvm::Constant *getConstantSignedPointer(CodeGenModule &CGM,
+                                         llvm::Constant *pointer,
```
Here and in many other places - use `PascalCase` for variables (https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly). The surrounding code uses the convention, so it looks like there is no reason not to follow it.

I'll leave a bunch of similar comments, but please do not limit your fixes of variable names to my comments since I might have missed some.

kovdan01 commented on clang/lib/CodeGen/CGExprConstant.cpp:
```diff
@@ -2061,6 +2071,58 @@ ConstantLValueEmitter::VisitCallExpr(const CallExpr *E) {
 }
 
+ConstantLValue
+ConstantLValueEmitter::emitPointerAuthSignConstant(const CallExpr *E) {
```
Names to fix: `unsignedPointer`, `key`, `storageAddress`, `otherDiscriminator`, `signedPointer`

kovdan01 commented on clang/lib/CodeGen/CGExprConstant.cpp:
```diff
@@ -2061,6 +2071,58 @@ ConstantLValueEmitter::VisitCallExpr(const CallExpr *E) {
 
+ConstantLValue
+ConstantLValueEmitter::emitPointerAuthSignConstant(const CallExpr *E) {
+  auto unsignedPointer = emitPointerAuthPointer(E->getArg(0));
```
For `unsignedPointer`, `key`, `signedPointer`: `auto` -> `auto *`

kovdan01 commented on clang/lib/CodeGen/CGExprConstant.cpp:
```diff
@@ -2061,6 +2071,58 @@ ConstantLValueEmitter::VisitCallExpr(const CallExpr *E) {
+  return signedPointer;
+}
+
+llvm::Constant *ConstantLValueEmitter::emitPointerAuthPointer(const Expr *E) {
```
Names to fix: `result`, `succeeded`

kovdan01 commented on clang/lib/CodeGen/CGExprConstant.cpp:
```diff
@@ -2061,6 +2071,58 @@ ConstantLValueEmitter::VisitCallExpr(const CallExpr *E) {
+}
+
+std::pair<llvm::Constant*, llvm::Constant*>
+ConstantLValueEmitter::emitPointerAuthDiscriminator(const Expr *E) {
```
Names to fix: `call`, `pointer`, `extra`, `result`

kovdan01 commented on clang/lib/CodeGen/CGExprConstant.cpp:
```diff
@@ -2061,6 +2071,58 @@ ConstantLValueEmitter::VisitCallExpr(const CallExpr *E) {
+ConstantLValueEmitter::emitPointerAuthDiscriminator(const Expr *E) {
+  E = E->IgnoreParens();
+
+  if (auto call = dyn_cast<CallExpr>(E)) {
```
For `call`, `pointer`, `extra`, `result`: `auto` -> `auto *`

kovdan01 commented on clang/lib/CodeGen/CGPointerAuth.cpp:
```diff
@@ -0,0 +1,77 @@
+//
+//===----------------------------------------------------------------------===//
+
+#include "CGCXXABI.h"
```
It looks like that only two includes are needed: CodeGenModule.h and clang/CodeGen/CodeGenABITypes.h. Maybe some other includes would become needed in future, but let's add them in corresponding PRs, not here.

kovdan01 commented on clang/lib/CodeGen/CGPointerAuth.cpp:
```diff
@@ -0,0 +1,77 @@
+
+/// Build a signed-pointer "ptrauth" constant.
+static llvm::ConstantPtrAuth *
+buildConstantAddress(CodeGenModule &CGM, llvm::Constant *pointer, unsigned key,
```
Naming style of variables here and all over the file

kovdan01 commented on clang/lib/CodeGen/CodeGenModule.h:
```diff
@@ -938,6 +938,11 @@ class CodeGenModule : public CodeGenTypeCache {
   // Return the function body address of the given function.
   llvm::Constant *GetFunctionStart(const ValueDecl *Decl);
 
+  llvm::Constant *getConstantSignedPointer(llvm::Constant *pointer,
```
Naming style of variables

kovdan01 commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2030,8 +2030,26 @@ bool Sema::checkConstantPointerAuthKey(Expr *Arg, unsigned &Result) {
+    return std::make_pair(nullptr, CharUnits());
+
+  // Base must be a declaration and can't be weakly imported.
+  auto baseDecl =
```
```suggestion
  const auto *BaseDecl =
```

kovdan01 commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2030,8 +2030,26 @@ bool Sema::checkConstantPointerAuthKey(Expr *Arg, unsigned &Result) {
 }
 
+static std::pair<const ValueDecl *, CharUnits>
+findConstantBaseAndOffset(Sema &S, Expr *E) {
```
Naming style of variables

kovdan01 commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2074,16 +2092,91 @@ static bool checkPointerAuthValue(Sema &S, Expr *&Arg,
+  // Must be a pointer or integer or blend thereof.
+  Expr *pointer = nullptr;
+  Expr *integer = nullptr;
+  if (auto call = dyn_cast<CallExpr>(Arg->IgnoreParens())) {
```
```suggestion
  if (auto *Call = dyn_cast<CallExpr>(Arg->IgnoreParens())) {
```

kovdan01 commented on clang/test/CodeGen/ptrauth-intrinsic-sign-constant.c:
```diff
@@ -0,0 +1,20 @@
+// RUN: %clang_cc1 -triple arm64-apple-ios -fptrauth-intrinsics -emit-llvm %s  -o - | FileCheck %s
```
It would be nice to also add a run line against `aarch64` or `aarch64-elf` generic triple in this and other tests. The logic tested is not apple-specific and locally I've ensured that adding this run line does not break tests.

asl replied:
@kovdan01 If you're having a RUN line locally, can you just suggest a change here?

kovdan01 replied:
> @kovdan01 If you're having a RUN line locally, can you just suggest a change here?

Updated the comment, thanks

kovdan01 reviewed (changes requested):
The changes themself in terms of functionality look OK to me - but I want someone else with deeper understanding of the context to take a look before this gets merged.

In terms of style, here are some of the categories of minor issues:
- unneeded includes
- wrong case for variables (need `PascalCase`, the surrounding code also uses that, so no need for not following the global convention https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly)
- missing `const` qualifiers and `auto` -> `auto *`

---

As for discussion above:

> Conceptually though, the distinction does seem useful, because we treat `__builtin_ptrauth_sign_unauthenticated` as dangerous (because in general we can't guarantee that it won't get lowered to a signing oracle), and only to be used carefully, when absolutely necessary (e.g., in a dynamic loader.) `__builtin_ptrauth_sign_constant` is safe, because we do always have safe ways to materialize signed constants (with the constants in IR, and either relocations or safe pseudos in the backend.)

I agree that having such a distinction seems useful.

asl commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2126,14 +2219,16 @@ static ExprResult PointerAuthSignGenericData(Sema &S, CallExpr *Call) {
   if (checkPointerAuthEnabled(S, Call))
     return ExprError();
-  if (checkPointerAuthValue(S, Call->getArgs()[0], OpKind) ||
+  if (checkPointerAuthValue(S, Call->getArgs()[0], OpKind, RequireConstant) ||
```
Can we check code like `Call->getArgs()[N]` to `Call->getArg(N)` while here? Certainly not everywhere, but on changed lines?

ahmedbougacha replied:
This one is actually interesting: we can't do that as-is, because the helper functions take an `Expr *&`, to do type conversion and placeholder expansion in-place;  see e.g., `convertArgumentToType`.
Another existing user makes the modification explicit using a local and `setArg`, but we have a lot of these calls in the checking code here.

We can rename the helper to something explicit though, perhaps `checkAndConvertPointerAuthValue`, and pass the call expr and arg index separately (same for key etc.)

asl replied:
Yes, `checkAndConvertPointerAuthValue` looks better to me as otherwise it is not pretty clear there are side effects.

ahmedbougacha replied:
Let's do that in a separate PR and adopt it for the other builtins there

ahmedbougacha committed:
Address feedback.

- test elf as well
- rename every variable.

efriedma-quic commented on clang/docs/PointerAuthentication.rst:
```diff
@@ -354,6 +354,23 @@ Given that ``signedPointer`` matches the layout for signed pointers signed with
+
+Calls to this are constant expressions if the discriminator is a null-pointer
+constant expression or an integer constant expression. Implementations may
+allow other pointer expressions as well.
```
This seems to imply that if you don't need a constant expression, the discriminator can be a variable.  This doesn't seem to match the implementation, though: it requires a constant discriminator in a specific form.

ahmedbougacha replied:
Yeah, I agree today this could simply be "it's always a constant expression";  I'll rewrite it (cc @rjmccall if this looks like anything to you)

kovdan01 commented on clang/lib/CodeGen/CGExprConstant.cpp:
```diff
@@ -1856,6 +1856,12 @@ class ConstantLValueEmitter : public ConstStmtVisitor<ConstantLValueEmitter,
+  ConstantLValue emitPointerAuthSignConstant(const CallExpr *E);
+  llvm::Constant *emitPointerAuthPointer(const Expr *E);
+  unsigned emitPointerAuthKey(const Expr *E);
+  std::pair<llvm::Constant*, llvm::Constant*>
```
From the function prototype, it's unclear that the first element of the pair stands for `StorageAddress` and the second one - for `OtherDiscriminator`. Consider adding a comment or, alternatively, using a named structure instead of `std::pair` (would break `std::tie` usage though)

ahmedbougacha replied:
The more interesting question is the usage of Constant;  I think we can safely assume the integer component is always going to be a ConstantInt (because it's going to have to fit in a ConstantInt-shaped hole in CPA anyway.)  That hopefully makes this obvious here, and I'll propagate that in the various users separately

kovdan01 commented on clang/lib/CodeGen/CGPointerAuth.cpp:
```diff
@@ -0,0 +1,63 @@
+buildConstantAddress(CodeGenModule &CGM, llvm::Constant *Pointer, unsigned Key,
+                     llvm::Constant *StorageAddress,
+                     llvm::Constant *OtherDiscriminator) {
+  llvm::Constant *AddressDiscriminator = nullptr;
```
Do we need to set this to `nullptr`? We 100% define this in either `if` or `else` branch and do not use the unitialized value before. Same applies to `IntegerDiscriminator` below.

You can also use smth like

```
llvm::Constant *AddressDiscriminator =
      (StorageAddress ? StorageAddress
                      : llvm::Constant::getNullValue(CGM.UnqualPtrTy));
  assert(!StorageAddress || (StorageAddress->getType() == CGM.UnqualPtrTy));
```

kovdan01 commented on clang/lib/CodeGen/CGExprConstant.cpp:
```diff
@@ -2061,6 +2071,58 @@ ConstantLValueEmitter::VisitCallExpr(const CallExpr *E) {
+ConstantLValueEmitter::emitPointerAuthDiscriminator(const Expr *E) {
+  E = E->IgnoreParens();
+
+  if (auto *Call = dyn_cast<CallExpr>(E)) {
```
```suggestion
  if (const auto *Call = dyn_cast<CallExpr>(E)) {
```

kovdan01 commented on clang/test/Sema/ptrauth.c:
```diff
@@ -58,6 +58,35 @@ void test_string_discriminator(const char *str) {
+  __builtin_ptrauth_sign_constant(&dv, mismatched_type, 0); // expected-error {{passing 'struct A' to parameter of incompatible type 'int'}}
+  __builtin_ptrauth_sign_constant(&dv, VALID_DATA_KEY, mismatched_type); // expected-error {{extra discriminator must have pointer or integer type; type here is 'struct A'}}
+
+  (void) __builtin_ptrauth_sign_constant(NULL, VALID_DATA_KEY, &dv); // expected-error {{argument to ptrauth_sign_constant must refer to a global variable or function}}
```
Is it a desired behavior that for `__builtin_ptrauth_auth` and `__builtin_ptrauth_sign_unauthenticated` we have special warning messages for null pointer passed as an argument, while here we have a generic error message? I'm OK with current state, I just want to be sure that this is the desired behavior.

ahmedbougacha replied:
We could special-case null pointers, but they're already covered by the diagnostic, which asks for global variables or functions – which NULL isn't.  For auth/sign, we don't have that sort of constraint on the pointer: it really is NULL and NULL alone that's special.

Now, the more interesting question is whether we should allow null pointers at all here.  Since defining these original builtins we have taught the qualifier to have a mode that signs/authenticates null, for some specific use-cases where replacing a signed value with NULL (which is otherwise never signed or authenticated) would bypass signing in a problematic way.
We haven't had the chance or need to revisit the builtins to allow sign/auth of NULL, but it's reasonable to add that support in the future.  We'd have to consider how to expose that in the builtins, because it's probably still something that's almost always a mistake;  more builtins would be an easy solution but maybe not a sophisticated one.

kovdan01 replied:
Thanks for explanation. I suppose it's OK to leave the way null sign/auth is handled "as is" for this PR and ship changes (if needed) as future patches.

kovdan01 commented on clang/lib/CodeGen/CGExprConstant.cpp:
```diff
@@ -2061,6 +2071,58 @@ ConstantLValueEmitter::VisitCallExpr(const CallExpr *E) {
+  unsigned Key = emitPointerAuthKey(E->getArg(1));
+  llvm::Constant *StorageAddress;
+  llvm::Constant *OtherDiscriminator;
+  std::tie(StorageAddress, OtherDiscriminator) =
```
Maybe just use a structured binding declaration like this?

```
  auto [StorageAddress, OtherDiscriminator] =
      emitPointerAuthDiscriminator(E->getArg(2));
```

Or do you want to have separate declarations with explicitly set types?

ahmedbougacha replied:
Yeah, this simply predates structured bindings;  we can indeed use them now.

kovdan01 commented on clang/docs/PointerAuthentication.rst:
```diff
@@ -354,6 +354,23 @@ Given that ``signedPointer`` matches the layout for signed pointers signed with
+a non-attackable sequence.
+
+``pointer`` must be a constant expression of pointer type which evaluates to
+a non-null pointer.  The result will have the same type as ``discriminator``.
```
`key` type is not described here (should be `ptrauth_key` as in comments in ptrauth.h)

kovdan01 commented on clang/docs/PointerAuthentication.rst:
```diff
@@ -354,6 +354,23 @@ Given that ``signedPointer`` matches the layout for signed pointers signed with
+  ptrauth_sign_constant(pointer, key, discriminator)
+
+Return a signed pointer for a constant address in a manner which guarantees
+a non-attackable sequence.
```
It would be nice to have this term described more deeply. Do you mean an absence of a signing oracle when talking about a non-attackable sequence?

ahmedbougacha replied:
Later additions to this document describe that in depth, you can look for
> [clang][docs] Document the ptrauth security model.

on my branch

kovdan01 replied:
Thanks, now I see that. Are you going to submit documentation changes all at once as a separate PR? If it's not too time-consuming, I'd prefer having all related terms defined as a part of this PR (particularly, at least the part "Attackable code sequences"), but I'm OK with submitting documentation changes separately.

kovdan01 commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2030,8 +2030,25 @@ bool Sema::checkConstantPointerAuthKey(Expr *Arg, unsigned &Result) {
+  if (!E->EvaluateAsRValue(Result, S.Context) || !Result.Val.isLValue())
+    return std::make_pair(nullptr, CharUnits());
+
+  // Base must be a declaration and can't be weakly imported.
```
Is there a test for this case?

ahmedbougacha replied:
I removed the weakref base pointer check because that's currently supported in the qualifier (and I think this should have also checked for weakimport as well for the cases I care about), so the backend has to deal with ptrauth weak refs anyway.  There's a higher-level set of problems around how we deal with weak refs in general, but that's a separate topic.

kovdan01 commented on clang/test/CodeGen/ptrauth-intrinsic-sign-constant.c:
```diff
@@ -0,0 +1,21 @@
+void *ptr2 = __builtin_ptrauth_sign_constant(&external, 2, __builtin_ptrauth_blend_discriminator(&ptr2, 26));
+
+// CHECK: @ptr3 = global ptr null
+void *ptr3;
```
Feel free to ignore: probably worth adding test for smth like

```
void *ptr4 = __builtin_ptrauth_sign_constant(&external, 0, NULL);
```

This doesn't bring that much value over testing a non-null constant integer discriminator, but:
1. In docs, having null pointer as a discriminator is listed as a separate option
   > This is a constant expression if the extra data is an integer or
   null pointer constant. 
2. With `NULL` defined as `#define NULL ((void*) 0)`, we'll explicitly test having a constant pointer, not an integer, value as a discriminator

ahmedbougacha replied:
(1) is an old artifact that I fixed separately, but (2) is interesting if we test real pointers, not necessarily just `NULL`;  added that and a few others

kovdan01 commented on clang/test/CodeGen/ptrauth-intrinsic-sign-constant.c:
```diff
@@ -0,0 +1,21 @@
+// CHECK: @ptr3 = global ptr null
+void *ptr3;
+
+void test_sign_constant_code() {
```
It's probably worth to have the exact same tests for global and function-local variables to ensure that everything is OK in both contexts.

kovdan01 commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2074,16 +2091,91 @@ static bool checkPointerAuthValue(Sema &S, Expr *&Arg,
+
+    // Must be rooted in a declaration reference.
+    if (!BaseOffsetPair.first) {
+      Invalid = true;
```
It's probably out of scope of this patch and may be just considered (or maybe not :) ) as a TODO for future enhancements - it's probably worth having separate error messages for different `Invalid = true` cases. This applies to both `err_ptrauth_bad_constant_discriminator` and `err_ptrauth_bad_constant_pointer`. Feel free to ignore though.

ahmedbougacha replied:
Heh yeah, we can probably add more in the future, but TBH this is obscure enough that we can do it if asked

kovdan01 commented on clang/lib/Sema/SemaChecking.cpp:
```diff
@@ -2074,16 +2091,91 @@ static bool checkPointerAuthValue(Sema &S, Expr *&Arg,
+    // never really be feasible and we'll have to catch it at link-time.
+    auto BaseOffsetPair = findConstantBaseAndOffset(S, Pointer);
+    if (!BaseOffsetPair.first || !isa<VarDecl>(BaseOffsetPair.first)) {
+      Invalid = true;
```
Is there a test for this case? I suppose that trying to use `__builtin_ptrauth_blend_discriminator` with a non-null pointer not actually pointing to some object should cover this - currently, I don't see such a test (please let me know if I missed it).

ahmedbougacha replied:
The missing base was covered, but the VarDecl check wasn't I think;  I added tests where the discriminator is a function decl

ahmedbougacha committed:
Simplify some functionality and address review feedback.

- don't reject weakref base pointers (allowed in quals now anyway)
- fold buildConstantAddress into getConstantSignedPointer (originally
  to provide a caching layer for global wrappers but not needed anymore)
- remove stripPointerCasts from it (originally for typed pointers)
- use ConstantInt for the integer/other component of discriminators
- remove CodeGen:: helper (originally for swift but not required)
- rejigger tests; try for parity in fn body and global init
- add a couple missing tests
- tweak doc/header comment
- adopt 'can be used in cst exprs' phrasing from LangExts doc
- various nits, naming, and format

ahmedbougacha committed:
Add more codegen tests.

ahmedbougacha commented on clang/lib/CodeGen/CodeGenModule.h:
```diff
@@ -938,6 +938,11 @@ class CodeGenModule : public CodeGenTypeCache {
+  llvm::Constant *getConstantSignedPointer(llvm::Constant *Pointer,
+                                           unsigned Key,
+                                           llvm::Constant *StorageAddress,
+                                           llvm::Constant *ExtraDiscrim);
```
Different people name these differently;  we can settle on OtherDiscriminator here yeah.  I think `extra`  as more of a common user-facing term for "discriminator", as in "extra data";  a lot of the PAC crowd coming from the security world thinks of it in those terms

kovdan01 commented on clang/docs/PointerAuthentication.rst:
```diff
@@ -354,6 +354,23 @@ Given that ``signedPointer`` matches the layout for signed pointers signed with
+a non-attackable sequence.
+
+``pointer`` must be a constant expression of pointer type which evaluates to
+a non-null pointer.  The result will have the same type as ``discriminator``.
```
Could you please reply to the previous comment in this thread? It was resolved, but the issue with the docs does not seem to be fixed. Please let me know if I miss smth and there is no issue - but, at least, we have different info in rst docs and in ptrauth.h header (the first one says that the result has the same type as `discriminator`, the second one - that the result's type is the same as the type of original value). So, things should at least be consistent. Ideally, we can just use the exact same text for description in both places, but I'm happy with different descriptions if they just match each other.

ahmedbougacha replied:
These two lines are just leftovers;  I deleted them in 1201821dc1e4.  The below is the same.

kovdan01 reviewed (approved):
LGTM with a couple of documentation-related comments non-blocking in terms of merge (can be fixed in follow-up PRs)

ahmedbougacha committed:
Remove leftover comment.
